core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/columns/DataColumnImpl.kt

```diff
@@ -1,27 +1,55 @@
 package org.jetbrains.kotlinx.dataframe.impl.columns
 
+import org.jetbrains.kotlinx.dataframe.ColumnDataHolder
 import org.jetbrains.kotlinx.dataframe.DataColumn
 import org.jetbrains.kotlinx.dataframe.api.dataFrameOf
+import org.jetbrains.kotlinx.dataframe.impl.isArray
+import org.jetbrains.kotlinx.dataframe.impl.isPrimitiveArray
+import kotlin.reflect.KClass
 import kotlin.reflect.KType
+import kotlin.reflect.full.isSubclassOf
+
+
+private const val DEBUG = true
 
 internal abstract class DataColumnImpl<T>(
-    protected val values: List<T>,
+    protected val values: ColumnDataHolder<T>,
     val name: String,
-    val type: KType,
-    distinct: Lazy<Set<T>>? = null
+    val type: KType
 ) : DataColumn<T>, DataColumnInternal<T> {
 
-    protected val distinct = distinct ?: lazy { values.toSet() }
+    private infix fun <T> T?.matches(type: KType) =
+        when {
+            this == null -> type.isMarkedNullable
+            this.isPrimitiveArray -> type.isPrimitiveArray &&
+                this!!::class.qualifiedName == type.classifier?.let { (it as KClass<*>).qualifiedName }
+
+            this.isArray -> type.isArray // cannot check the precise type of array
+            else -> this!!::class.isSubclassOf(type.classifier as KClass<*>)
+        }
+
+    init {
+        if (DEBUG) {
```

Comment:
![4e4e008a64d0a6fe0100b96eabdd48cd](https://github.com/Kotlin/dataframe/assets/1198621/ec68be38-07c5-45bd-b829-ba4a10ff3bed)
We are close to the Android macroses!

Reply:
hahah, yeah it was just for research :)